pkg/operator/ceph/cluster/cluster.go

```diff
@@ -397,6 +411,86 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
 	return nil
 }
 
+func extractExitCode(err error) (int, bool) {
+	exitErr, ok := err.(*exec.ExitError)
+	if ok {
+		return exitErr.ExitCode(), true
+	}
+	return 0, false
+}
+
+func (c *cluster) removeDefaultReplicationRule() error {
+	args := []string{"osd", "crush", "rule", "rm", "replicated_rule"}
+	cephCmd := client.NewCephCommand(c.context, c.Namespace, args)
+	_, err := cephCmd.Run()
+	if err != nil {
+		if code, ok := extractExitCode(err); ok && code == int(syscall.EBUSY) {
+			// we do not want to delete the replicated_rule if it’s in use,
+			// and we also do not care much. There are two possible causes:
+			// - the user has created this rule with the non-default CRUSH
+			//   root manually
+			// - the user is using this rule despite the rule using the default
+			//   CRUSH root
+			// in both cases, we cannot do anything about it either way and
+			// we’ll assume that the user knows what they’re doing.
+			logger.Debug("replicated_rule is in use, not removed")
+			return nil
+		}
+		// the error does not refer to EBUSY -> return as error
+		return errors.Wrap(err, "failed to remove default replicated_rule")
+	}
+	return nil
+}
+
+func (c *cluster) removeDefaultCrushRoot() error {
+	args := []string{"osd", "crush", "rm", "default"}
+	cephCmd := client.NewCephCommand(c.context, c.Namespace, args)
+	_, err := cephCmd.Run()
+	if err != nil {
+		if code, ok := extractExitCode(err); ok {
+			if code == int(syscall.ENOTEMPTY) || code == int(syscall.EBUSY) {
+				// we do not want to delete the default node if it’s in use,
+				// and we also do not care much. There are two more causes here:
+				// - a (non-root?) CRUSH node with the default label was created
+				//   automatically, e.g. from topology labels, and OSDs (or sub
+				//   nodes) have been placed in there. In this case, the node
+				//   obviously needs to be preserved.
+				// - the root=default CRUSH node is in use by a non-default
+				//   CRUSH rule
+				// - OSDs or subnodes have been placed under the root=default
+				//   CRUSH node
+				//
+				// in all cases, we cannot do anything about it either way and
+				// we’ll assume that the user knows what they’re doing.
+				logger.Debug("default is not empty or is still in use, not removed")
+				return nil
+			}
+		}
+		// the error does not refer to EBUSY or ENOTEMPTY -> return as error
+		return errors.Wrap(err, "failed to remove default CRUSH root")
+	}
+	return nil
+}
+
+// Remove the default root=default and replicated_rule CRUSH objects which are created by Ceph on initial startup.
+// Those objects may interfere with the normal operation of the cluster.
+// Note that errors which indicate that the objects are in use are ignored and the objects will continue to exist in that case.
+func (c *cluster) removeDefaultCrushMap() (err error) {
+	logger.Info("removing default objects, if applicable")
+
+	err = c.removeDefaultReplicationRule()
+	if err != nil {
+		return err
```

Comment:
@leseb Hm, is that a policy?

The errors returned by the methods are already wrapped, and I can’t think of any string for use with errors.Wrap which isn’t redundant:

- "failed to remove default CRUSH map" is redundant with the errors.Wrap which is already happening at the call site in postMonStartupActions
- "failed to remove default replication_rule rule" is redundant with the wrapping in removeDefaultReplicationRule

Maybe I’m using this wrong?

Reply:
I know that in some case it's redundant but this is a practice we are trying to enforce since it makes debugging simpler. 

What about `failed to remove default rule`? and `removeDefaultReplicationRule()` is more precise since it names the rule.